vochain/state/sik.go

```diff
@@ -0,0 +1,128 @@
+package state
+
+import (
+	"encoding/hex"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/tree/arbo"
+)
+
+// encodedHeightLen constant is the number of bytes of the encoded hysteresis
+// that contains the hysteresis height value, starting from the last byte
+const encodedHeightLen = 4
+
+// hysteresisLen constant contains the number of bytes that the encode
+// hysteresis has
+const hysteresisLen = 32
+
+// SetSIK function creates or update the SIK of the provided address in the
+// state. It covers the following cases:
+//   - It checks if already exists a valid SIK for the provided address and if
+//     so it returns an error.
+//   - If no SIK exists for the provided address, it will create one with the
+//     value provided.
+//   - If exists a logically deleted SIK for the address provided, checks the
+//     hysteresis height for that address before update its SIK.
+//   - If the hysteresis height is not reached yet, it returns an error.
+//   - If the hysteresis height is reached, it updates the value of the sik with
+//     the provided one.
+func (v *State) SetSIK(address common.Address, newSik []byte) error {
```

Comment:
I think SetSik should just set the sik. All the logic remaining hysteresis and the conditions to reset an existing sik should be part of the APP logic, not the state. Specifically part of the transaction handler.